### PR TITLE
arc: add STACK_SENTINEL and improve exception handling

### PR DESCRIPTION
**Pull request by vonhust**
This PR is consisted of three parts.

# PART I:  optimize the exception handling
   The original exception handling has space to optimize and
and some bugs need to be fixed.

* define NANO_ESF
   * add the definition of NANO_ESF which is an irq_stack_frame
   * add the corresponding codes in exception entry and handler
   * remove _default_esf
* implement the _ARCH_EXCEPT
   * use trap exception to raise exception by kernel
   * add corresponding trap exception entry
   * add _do_kernel_oops to handle the exception raised by
     _ARCH_EXCEPT.
* add the thread context switch in exception return
   * case: kernel oops may raise thread context switch
   * case: some tests will re-implement SysFatalHandler to raise
     thread context switch.
   * as the exception and isr are handled in kernel isr stack, so
     the thread context switch must be in the return of exception/isr
     , and the exception handler must return, should not be decorated
     with FUNC_NORETURN
* for arc, _is_in_isr should consider the case of exception
* bug fixes

# PART II: add the support of STACK_SENTINEL
  * call the _check_stack_sentinel in unnested isr
return.

# PART III: necessary changes in tests and kernel codes
   * code fixes for arc related tests
   * add CODE_UNREACHABLE in _StackCheckHandler

The three parts are a whole to pass the tests/kernel/fatal

The issue #6513 and #7055 can be fixed with this PR.

### CODE REVIEW

**Review by tyrantlizard77 — COMMENTED** [review]
Why do we need stack sentinel support on ARC?
It has a hardware feature to detect stack overflows, CONFIG_ARC_STACK_CHECKING

**Comment by vonhust** [comment]
@andrewboie , the STACK_CHECKING is an optional hardware feature of ARC, not a default configuration.
If the arc-based processors are not configured with STACK_CHECKING and MPU,  STACK_SENTINEL is one software option to do stack overflow check.

File: arch/arc/core/fault.c [hunk 1]
```diff
@@ -29,7 +29,7 @@
  *
  * @return This function does not return.
```
Comment (ruuddw) [inline arch/arc/core/fault.c]: Please update the 'this function does not return' comment, the function could return for stack canarie exceptions, right?

**Review by ruuddw — APPROVED** [review]
Changes look sound to me, but for this change to exception handling I do recommend some extra testing of various configurations (with/without userspace, stack protection variants, etc.)

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**